models/currency.js

```diff
@@ -0,0 +1,189 @@
+      [currencyData.src]: {
+        [currencyData.target]: currencyData.quantity
+      },
+      lastUpdate: new Date()
+    }, {
+      merge: true
+    })
+    return true
+  } catch (err) {
+    logger.error('Error while adding currency rates into exchange', err)
+    throw err
+  }
+}
+
+/**
+ * Fetch the all banks availabe name and id
+ * @return {Promise<exchangeData>}
+ */
+const getAllBanks = async () => {
+  try {
+    const bankSnapshot = await bankModel.get()
+    const bankData = []
+    bankSnapshot.forEach(bank => {
+      const data = bank.data()
+      if (data.isActive) {
+        bankData.push({ bankId: data.bankId, bankName: data.bankName })
+      }
+    })
+    return bankData
+  } catch (err) {
+    logger.error('Error while getting bank detais', err)
+    throw err
+  }
+}
+
+/**
+ * Extracts the data form the first doc of firestore response
+ * @return {Promise<object>}
+ */
+const extractRefDocsData = (doc) => {
+  return doc.docs[0] && doc.docs[0].data()
+}
+
+/**
+ * Gets the id of frestore documenr
+ * @return {Promise<object>}
+ */
+const extractRefDocsId = (doc) => {
+  return doc.docs[0] && doc.docs[0].id
+}
+
+/**
+ * Update bank currecy after excbange
+ * @return {Promise<exchangeData>}
+ */
+const updateBankWalletAfterExchange = (bankCurrency, exchangeData, totalTargetCurrencyRequest) => {
+  bankCurrency[exchangeData.src] = (bankCurrency[exchangeData.src] || 0) + parseInt(exchangeData.quantity)
+  bankCurrency[exchangeData.target] -= totalTargetCurrencyRequest
+  return { ...bankCurrency }
+}
+
+/**
+ * Update user currecy after excbange
+ * @return {Promise<exchangeData>}
+ */
+const updateUserWalletAfterExchange = (userCurrency, exchangeData, totalTargetCurrencyRequest) => {
+  userCurrency[exchangeData.target] = (userCurrency[exchangeData.target] || 0) + parseInt(totalTargetCurrencyRequest)
+  userCurrency[exchangeData.src] -= exchangeData.quantity
+  return { ...userCurrency }
+}
+
+/**
+ * Transaction for currency exchange
+ * @return {Promise<exchangeData>}
+ */
+const exchangeTransaction = async (userId, exchangeData) => {
+  try {
+    const { bankId } = exchangeData
+    // db models refs
+    const userWalletRef = await walletModel.where('userId', '==', userId).limit(1)
+    const bankWalletRef = await bankModel.where('bankId', '==', bankId).limit(1)
+    const exchangeRateRef = await currencyExchangeModel.doc(allCurrencyName)
+    const exchangeResponse = await firestore.runTransaction(async t => {
+      const responseObj = {
+        status: false,
+        message: ''
+      }
+
+      const userWalletDoc = await t.get(userWalletRef)
+      const userWalletId = await extractRefDocsId(userWalletDoc)
+      const userWallet = await extractRefDocsData(userWalletDoc)
+      const bankWalletDoc = await t.get(bankWalletRef)
+      const bankWalletId = await extractRefDocsId(bankWalletDoc)
+      const bankWallet = await extractRefDocsData(bankWalletDoc)
```

Comment:
I guess all of these calls will wait for every previous call to complete, slowing down the process (correct me if I'm wrong)

I guess that the `2nd and 3rd` and `5th and 6th` calls can be solved using `Promise.all` . What do you say? I don't have that good understanding of promises, so let me know if I'm saying anything wrong 😄

Reply:
All read and writes are taking place in transaction